### DISCUSSION TITLE
Message Loss

### DISCUSSION THREAD

**Discussion by nshimizups** [microsoft/msquic#4715] [Q&A]
**Program Overview:**

- A program designed to perform 100 consecutive StreamSend operations.
- After executing a StreamSend using the `Send` function, the `Sending` function sets `m_isSendComplete` to `false`.
- `m_isSendComplete` is updated to `true` upon detection of the `QUIC_STREAM_EVENT_SEND_COMPLETE` callback.
- The program loops and waits while monitoring for the `QUIC_STREAM_EVENT_SEND_COMPLETE` callback.

**Results:**

- All 100 StreamSend operations returned normal status from `MsQuic->StreamSend`.
- Data loss has been observed on the receiving side, resulting in the total received data size being less than the total transmitted data size.

**Issues:**

- Data loss occurs on the receiving side.

**Question:**

- Introducing `sleep()` within the loop has reduced the occurrence of data loss. However, adding arbitrary sleep is not preferred.
- Are there any appropriate callback events or mechanisms that could be used as better判断基準 for when to perform the next data transmission?

**Test Program Excerpt (の一部抜粋):**

- In the `MainLoop` function, `MsQuic::StreamStart` has been completed before execution.

```c++
void TestApp::MainLoop(){

	char msg[16040] = {""""};

	while(true){

		if(false == isSendComp()){
			continue;
		}
		Send(static_cast<const void*>(msg),sizeof(msg));
		Sending();

		i++;
		if(i >= 100 ){
			break;
		}
	}
}

int TestApp::Send(const void *pBuff, int nBuffSize){

	// ... (previous code)...

	if (QUIC_FAILED(m_nErrno = MsQuic->StreamSend(m_Stream, SendBuffer, 1, QUIC_SEND_FLAG_NONE, NULL))) {
		return -1;
	}
	
	return 0;
}

void CTestApp::Sending()
{
	m_isSendComplete = false;
}

bool CTestApp::isSendComp()
{
	return m_isSendComplete;
}

**Comment by nibanks** [comment]
Getting a send complete only means the peer transport later has received the data, not that the app has. So, you cannot use that signal to determine if the peer got your data.

**Reply by nshimizups** [reply]
Thank you very much. 
So, how can we ensure that there is no loss of messages?

**Comment by nshimizups** [comment]
After trying various things, we were able to send and receive messages without any loss.

What we did:

We ensured not to close the stream in the QUIC_STREAM_EVENT_SEND_COMPLETE event.
When receiving notifications, we checked whether the data was stored beyond Event->RECEIVE.Buffers[1] according to Event->RECEIVE.BufferCount.
